Add provider for MVT tile generation from Postgres

ThorodanBrom opened the pull request:
# Overview

This PR adds MVT tile generation from Postgres/PostGIS tables natively.  

# Related Issue / discussion

This is related to issue #1978 

<!--

Is there an existing Issue that this PR addresses?  Does this PR need a new Issue?

Non-trivial PRs are best put forth initially as an Issue so that there can be
discussion and consensus before a PR is put forth.

-->

# Additional information

# Dependency policy (RFC2)

- [x] I have ensured that this PR meets [RFC2](https://pygeoapi.io/development/rfc/2) requirements

# Updates to public demo

- [x] I have ensured that breaking changes to the [pygeoapi master demo server](https://github.com/geopython/demo.pygeoapi.io) have been addressed
  - [x] https://github.com/geopython/demo.pygeoapi.io/blob/master/services/pygeoapi_master/local.config.yml

# Contributions and licensing

(as per https://github.com/geopython/pygeoapi/blob/master/CONTRIBUTING.md#contributions-and-licensing)

- [x] I'd like to contribute this feature to pygeoapi. I confirm that my contributions to pygeoapi will be compatible with the pygeoapi license guidelines at the time of contribution
- [x] I have already previously agreed to the pygeoapi Contributions and Licensing Guidelines

doublebyte1 reviewed (commented):
Thank You @ThorodanBrom ! Would you be able to add a Postgres Provider docker example at pygeoapi-examples? https://github.com/geopython/pygeoapi-examples/tree/main/docker

doublebyte1 commented on pygeoapi/provider/mvt_postgres.py:
```diff
@@ -0,0 +1,263 @@
+            query = text("""
+                WITH                                                                  
+                    bounds AS (
+                        SELECT ST_TileEnvelope(:z, :y, :x, ST_MakeEnvelope(-180, -90, 180, 90, 4326)) AS boundgeom
```
Why the tile envelope here is based on the extent of the crs, instead of the extent of the geometry (like in the case of WebMercator)?

ThorodanBrom replied:
[ST_TileEnvelope](https://postgis.net/docs/ST_TileEnvelope.html) by default is configured to EPSG:3857. In order to use it for other CRSs, the docs mention that you can add an extra parameter of a geometry that has an extent of the 0th zoom level. There's an example of setting it for EPSG:4326, we just used that here. (hope that this answered your question, if not, please let me know)

We can probably use the WKB representation there so that `MakeEnvelope` isn't calculated for every tile request.

doublebyte1 replied:
I agree, I think the WKB representation would make sense here!

ThorodanBrom replied:
Addressed in [97ba024](https://github.com/geopython/pygeoapi/pull/1979/commits/97ba024f4aa8a253ff04ac718bfb3feba528080a). I used EWKT representation instead of WKB as the the WKB hex string was too long

doublebyte1 commented on pygeoapi/provider/mvt_postgres.py:
```diff
@@ -0,0 +1,263 @@
+                        SELECT ST_TileEnvelope(:z, :y, :x) AS boundgeom
+                    ),
+                    mvtgeom AS (
+                        SELECT ST_AsMVTGeom(ST_Transform({geom}, 3857), bounds.boundgeom) AS geom {fields}
```
I think converting curves to lines (ST_CurveToLine), could speed up queries in the case of more complex geometries.

ThorodanBrom replied:
OK, I will try this out, [it looks like Martin also uses it](https://maplibre.org/martin/sources-pg-functions.html#simple-function). 

So by using `ST_CurveToLine`, will this provider also be able to support `CIRCULARSTRING`-like data types (that is what I understood from reading the [documentation](https://postgis.net/docs/ST_CurveToLine.html))? Or is it just going to be used for query performance?

doublebyte1 replied:
I think simplification, in general, yields performance; and in the case of tiles, we do not need the geometries to be so detailed.

doublebyte1 commented:
@ThorodanBrom could you check the documentation warnings?

doublebyte1 commented:
* If possible, please also add a test for this provider. You can check this one for inspiration: https://github.com/geopython/pygeoapi/blob/master/tests/test_postgresql_provider.py
* Please check the errors when building the documentation.

ThorodanBrom commented:
Thanks for the review,
- I'll work on adding an example provider for this in `pygeoapi-examples` as well as writing test cases
- I'll check out the errors in the documentation and flake8
- The PR needs a few updates and also needs to be rebased as well to accomodate the MVT tile fixes, is it OK if I rebase and force-push or will that break any of the CI flows?

doublebyte1 commented:
> Thanks for the review,
> 
> * I'll work on adding an example provider for this in `pygeoapi-examples` as well as writing test cases
> * I'll check out the errors in the documentation and flake8
> * The PR needs a few updates and also needs to be rebased as well to accomodate the MVT tile fixes, is it OK if I rebase and force-push or will that break any of the CI flows?

Thank You @ThorodanBrom ! I think it is ok to rebase, but I am pinging @tomkralidis here, just in case.

tomkralidis commented:
+1 to rebase, given it will start CI again fresh.

doublebyte1 commented on docs/source/data-publishing/ogcapi-tiles.rst:
```diff
@@ -130,6 +131,46 @@ Following code block shows how to configure pygeoapi to read Mapbox vector tiles
+           table: hotosm_bdi_waterways
+           geom_field: foo_geom
+           options:
+           zoom:
```
@ThorodanBrom I think that the `zoom` should be placed within `options`

doublebyte1 reviewed (changes requested):
Please fix the format of the example in the documentation, and I think we are good to go. 

You can also click on the check-boxes that say "Updates to public demo", as this provider is not used in the configuration file of the demo.

Thank You!

ThorodanBrom commented:
Thanks for the review @doublebyte1. I'm working on the example in `pygeoapi-examples`, but I'm not really sure on how to write tests for this provider. There aren't any tests for other MVT providers, so I don't have a reference. I can think of the following conditions to test:
- if the provider is able to create tiles for the configured table and configured zoom levels for `WebMercatorQuad` and `WorldCRS84Quad`
- if the tiles returned are valid MVT files + a `200 OK` 
    - I don't know how to test for a 'valid' MVT, is there a library or something that can parse and validate the API output 
- if tiles requested for an invalid index or where data does not exist returns a `204 No Content`

Do let me know if this sounds OK.

doublebyte1 commented:
> Thanks for the review @doublebyte1. I'm working on the example in `pygeoapi-examples`, but I'm not really sure on how to write tests for this provider. There aren't any tests for other MVT providers, so I don't have a reference. I can think of the following conditions to test:
> 
> * if the provider is able to create tiles for the configured table and configured zoom levels for `WebMercatorQuad` and `WorldCRS84Quad`
> * if the tiles returned are valid MVT files + a `200 OK`
>   
>   * I don't know how to test for a 'valid' MVT, is there a library or something that can parse and validate the API output
> * if tiles requested for an invalid index or where data does not exist returns a `204 No Content`
> 
> Do let me know if this sounds OK.

You are right in that there are no tests for other MVT providers. Let's keep it like this, for now. Thank You for the availability!

tomkralidis commented on docs/source/data-publishing/ogcapi-tiles.rst:
```diff
@@ -22,6 +22,7 @@ pygeoapi core tile providers are listed below, along with supported features.
    `MVT-elastic`_,✅,✅,✅,❌,❌,✅
    `MVT-proxy`_,❓,❓,❓,❓,❌,✅
    `WMTSFacade`_,✅,❌,✅,✅,✅,❌
+   `MVT-postgres`_,✅,✅,✅,✅,❌,✅
```
Rename to `MVT-postgresql`

tomkralidis commented on docs/source/data-publishing/ogcapi-tiles.rst:
```diff
@@ -130,6 +131,46 @@ Following code block shows how to configure pygeoapi to read Mapbox vector tiles
+.. note::
+   Requires Python packages sqlalchemy, geoalchemy2 and psycopg2-binary
+
+Must have PostGIS installed with protobuf-c support. 
```
Turn into an additional note

tomkralidis commented on docs/source/data-publishing/ogcapi-tiles.rst:
```diff
@@ -130,6 +131,46 @@ Following code block shows how to configure pygeoapi to read Mapbox vector tiles
+.. note::
+   Geometry must be using EPSG:4326
+
+This provider gives support to serving tiles generated using `Postgres <https://www.postgresql.org/>`_ with `PostGIS <https://postgis.net/>`_.
```
`s/Postgres/PostgreSQL/g`

tomkralidis commented on docs/source/data-publishing/ogcapi-tiles.rst:
```diff
@@ -130,6 +131,46 @@ Following code block shows how to configure pygeoapi to read Mapbox vector tiles
+
+   providers:
+       - type: tile
+           name: MVT-postgres
```
`MVT-postgresql`

tomkralidis commented on pygeoapi/plugin.py:
```diff
@@ -53,6 +53,7 @@
         'MVT-tippecanoe': 'pygeoapi.provider.mvt_tippecanoe.MVTTippecanoeProvider',  # noqa: E501
         'MVT-elastic': 'pygeoapi.provider.mvt_elastic.MVTElasticProvider',
         'MVT-proxy': 'pygeoapi.provider.mvt_proxy.MVTProxyProvider',
+        'MVT-postgres': 'pygeoapi.provider.mvt_postgres.MVTPostgresProvider',
```
`'MVT-postgresql': 'pygeoapi.provider.mvt_postgresql.MVTPostgreSQLProvider',`

tomkralidis commented on pygeoapi/provider/mvt_postgres.py:
Rename to `pygeoapi/provider/mvt_postgresql.py`

tomkralidis commented on pygeoapi/provider/mvt_postgres.py:
```diff
@@ -0,0 +1,246 @@
+# =================================================================
+#
+# Authors:
```
Please add yourself as author.  Example in https://github.com/geopython/pygeoapi/blob/master/setup.py#L3

tomkralidis commented on pygeoapi/provider/mvt_postgres.py:
```diff
@@ -0,0 +1,246 @@
+#
+# Authors:
+#
+# Copyright (c)
```
Please add copyright.  Example in https://github.com/geopython/pygeoapi/blob/master/setup.py#L5

tomkralidis commented on pygeoapi/provider/mvt_postgres.py:
```diff
@@ -0,0 +1,246 @@
+#
+# =================================================================
+
+import logging
```
Order imports by standard library, 3rd party packages, then local imports, alphabetically, separated by blank line.

```python
from copy import deepcopy
import logging

from sqlalchemy.sql import text

from pygeoapi.models.provider.base import (
    TileSetMetadata, TileMatrixSetEnum, LinkType)
from pygeoapi.provider.base import ProviderConnectionError
from pygeoapi.provider.base_mvt import BaseMVTProvider
from pygeoapi.provider.postgresql import PostgreSQLProvider
from pygeoapi.util import url_join
```

tomkralidis commented on pygeoapi/provider/mvt_postgres.py:
```diff
@@ -0,0 +1,246 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class MVTPostgresProvider(BaseMVTProvider):
```
Rename to `MVTPostgreSQLProvider`

tomkralidis reviewed (changes requested):
Great work here!  Minor change requests on naming consistency.

ThorodanBrom commented:
Hello @tomkralidis, thanks for the review. I think I've addressed all your comments in 9a320cc. (I added a small bug fix in 
a034d15 as well). Please let me know if there are any other changes required.

photonbit commented on docs/source/data-publishing/ogcapi-tiles.rst:
```diff
@@ -130,6 +131,47 @@ Following code block shows how to configure pygeoapi to read Mapbox vector tiles
+   Must have PostGIS installed with protobuf-c support 
+
+.. note::
+   Geometry must be using EPSG:4326
```
I understand that this note is outdated for the postgresql provider. Is the note valid for this case?
